src/Flysystem/FlysystemAssetStore.php

```diff
@@ -1488,16 +1498,38 @@ public function getResponseFor($asset)
             $code = $redirectFileID === $permanentFileID ?
                 $this->config()->get('permanent_redirect_response_code') :
                 $this->config()->get('redirect_response_code');
-            return $this->createRedirectResponse($redirectFileID, $code);
+
+            return $this->extendResponse(
+                $this->createRedirectResponse($redirectFileID, $code),
+                $asset,
+                ['visibility' => self::VISIBILITY_PUBLIC, 'parsedFileID' => $parsedFileID]
+            );
         }
 
         // Deny if file is protected and denied
         if ($protected->has($asset)) {
-            return $this->createDeniedResponse();
+            return $this->extendResponse(
+                $this->createDeniedResponse(),
+                $asset,
+                ['visibility' => self::VISIBILITY_PROTECTED]
+            );
         }
 
         // We've looked everywhere and couldn't find a file
-        return $this->createMissingResponse();
+        return $this->extendResponse($this->createMissingResponse(), $asset);
+    }
+
+    /**
+     * Wrap an "updateResponse" extend call around the response along with some extra context.
+     * @param HTTPResponse $response Response that could be updated
+     * @param string $asset Original file id of the request
+     * @param array $context Additional context to send along
+     * @return HTTPResponse Updated response
+     */
+    private function extendResponse(HTTPResponse $response, string $asset, array $context = [])
+    {
+        $this->extend('updateResponse', $response, $asset, $context);
```

Comment:
Context is just meant to provide extra details about the request and how the response was built. At this stage it could contain a `visibility` attribute to say what store the file was found on and a `parsedFileID`.